Add reserved descriptor ID range

mrtracy opened the pull request:
This commit splits the existing range of "reserved" descriptor IDs into two
sets; there is now a range of "System" IDs to hold SystemDB and its tables, and
another non-system "Reserved" range which will hold tables used internally by
CockroachDB that should not be part of SystemDB.

This is being done to support upcoming log-structured tables that Cockroach will
be writing to internally; these tables must be present at cluster bootstrap, but
should also not have their contents gossiped to all nodes.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3425)

<!-- Reviewable:end -->

Commit:
Add reserved descriptor ID range

This commit splits the existing range of "reserved" descriptor IDs into two
sets; there is now a range of "System" IDs to hold SystemDB and its tables, and
another non-system "Reserved" range which will hold tables used internally by
CockroachDB that should not be part of SystemDB.

This is being done to support upcoming log-structured tables that Cockroach will
be writing to internally; these tables must be present at cluster bootstrap, but
should also not have their contents gossiped to all nodes.

tamird commented:
Why do non-gossiped tables need reserved IDs?

mrtracy commented:
When I initially refactored the initial schema creation, non-system tables were allocated the same descriptors as User tables.  However, I backed off of that for a couple of reasons:

1. This broke assumptions in a large number of tests about the number of descriptors present at startup. After thinking about it myself (and consulting with others), we decided that it was fair for tests to assume that no other "User" tables would be present at startup, and that another reserved range was an appropriate way to ensure that.
2. These tables are not gossiped, but they will likely get some special semantics of their own; for instance, it should eventually be impossible for the root user to DROP these tables.  A reserved descriptor ID range is an easy way to allow these semantics.

tamird commented:
> 1. This broke assumptions in a large number of tests about the number of descriptors present at startup. After thinking about it myself (and consulting with others), we decided that it was fair for tests to assume that no other "User" tables would be present at startup, and that another reserved range was an appropriate way to ensure that.

Seems like a reasonable hack in the short term, but we should file an issue and clean this up.

> 2. These tables are not gossiped, but they will likely get some special semantics of their own; for instance, it should eventually be impossible for the root user to DROP these tables. A reserved descriptor ID range is an easy way to allow these semantics.

Same comment as above. An accumulation of meanings for table IDs is my main concern here.

mrtracy commented:
I've logged #3444 to address your concerns and clean this up.

As for this current solution, is the code acceptable?

petermattis commented:
Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[config/config.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YMOpF12qU6S1SV16_:-K5YMOpF12qU6S1SV16a:107978037) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/config/config.go#L150)):*
This would be clearer if you simply tried to decode the varint tableID at the start of key. You'd also want to make sure the key starts with a varint encoding:

```
  if encoding.PeekType(key) != encoding.Int {
    return 0, fmt.Errorf(...)
  }
  key, tableID, err := encoding.DecodeUvarint(key)
  if err != nil {
    return 0, err
  }
  if tableID != keys.DescriptorTableID {
    return 0, fmt.Errorf(...)
  }
```

Even better if you moved part of this into a `keys.DecodeTablePrefix` function.

---

*[keys/constants.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YNXS6MXMEGItrkboN:-K5YNXS6MXMEGItrkboO:-704796503) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/keys/constants.go#L194)):*
I don't feel there is a clear distinction between system tables and reserved tables. I'd prefer an approach where all internal tables are called "system" tables but we somehow annotate the ones which need to be gossiped or other special processing. 

---

*[keys/constants.go, line 202 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YNnzTIs44K3uw1jh4:-K5YNnzTIs44K3uw1jh5:-968203758) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/keys/constants.go#L202)):*
`s/.././g`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
There may be alternative solutions to the DROP permissions without reserving an ID range, but one thing the reserved ID range gives us that would be more difficult with an alternative strategy is the ability to cover all the system tables with a single zone config.  

LGTM

---

Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/constants.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YNXS6MXMEGItrkboN:-K5aOGqmWlufBolX4w0x:-1954908310) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/keys/constants.go#L194)):*
One reason we reserve IDs for gossiped system tables is so that they are all contiguous and can be loaded efficiently for gossip. I think we should keep a special reserved range for the gossiped tables, instead of mixing them with non-gossiped tables with some sort of annotation.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-:-K5aP_H1E0HbFxOcMm87:-1545246032)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[config/config.go, line 150 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YMOpF12qU6S1SV16_:-K5b8SllRHtlVdlurJ6e:1162594601) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/config/config.go#L150)):*
Done, added a keys.DecodeTablePrefix which makes this a little clearer.

---

*[keys/constants.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YNXS6MXMEGItrkboN:-K5b1BFmBpH-rN6Y8cBu:-1887788975) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/keys/constants.go#L194)):*
I've added a big TODO at this location which also links to the github issue.  We've agreed to go with this solution for now, but I don't think we'll forget to come back and find a solution with broader support.

---

*[keys/constants.go, line 202 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YNnzTIs44K3uw1jh4:-K5b24CIw3RObH096j3S:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/keys/constants.go#L202)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 9 of 11 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[config/config.go, line 154 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5etsQgWvuSBaOWfbNO:-K5etsQgWvuSBaOWfbNP:-1539097419) ([raw file](https://github.com/cockroachdb/cockroach/blob/38143bc69aff6f6515a306657251cdd64806a413/config/config.go#L154)):*
should be util.Errorf

---

*[config/config.go, line 181 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5eu9uUuWl6-3oSPFY4:-K5eu9uUuWl6-3oSPFY5:1691631138) ([raw file](https://github.com/cockroachdb/cockroach/blob/38143bc69aff6f6515a306657251cdd64806a413/config/config.go#L181)):*
drive-by, but remove this; the error below gives the same information

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[keys/constants.go, line 194 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5YNXS6MXMEGItrkboN:-K5ez9kNo3q0BWtM2OXS:1476809963) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea65fc18b04dffbfdd1b14d8ef413ed2d314784a/keys/constants.go#L194)):*
I agree that the gossiped tables should be contiguous for easy gossiping. But we're already gossiping more tables and data than we need to and I'm struggling to imagine when we'd add more. I would be fine with a scheme that reserved some very small number of system IDs for gossiping (e.g. the first 8) and then reserved a larger number (64) for other system tables (e.g. leases, users, zones, event-log, etc). 

---

*[keys/keys.go, line 348 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5ezqiNulPScRU3tw49:-K5ezqiNulPScRU3tw4A:2051383637) ([raw file](https://github.com/cockroachdb/cockroach/blob/38143bc69aff6f6515a306657251cdd64806a413/keys/keys.go#L348)):*
If you feeling like being a little OCD: the error message we've used elsewhere is `"invalid key prefix: %q"`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-:-K5f-3y_GW15_nzsD8xp:676834233)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[config/config.go, line 154 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5etsQgWvuSBaOWfbNO:-K5gY0rB0--oGy1O6KZ6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/38143bc69aff6f6515a306657251cdd64806a413/config/config.go#L154)):*
Done.

---

*[config/config.go, line 181 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5eu9uUuWl6-3oSPFY4:-K5gY19wv7DZhm3MpuF2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/38143bc69aff6f6515a306657251cdd64806a413/config/config.go#L181)):*
Done.

---

*[keys/keys.go, line 348 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425#-K5ezqiNulPScRU3tw49:-K5gXWiLK7nK0bdeZpfL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/38143bc69aff6f6515a306657251cdd64806a413/keys/keys.go#L348)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3425)*
<!-- Sent from Reviewable.io -->
